### PR TITLE
ROC-5998 Add generic endpoint for pdf generation / retrieval from dm store + refactor

### PR DESCRIPTION
**Pull request by simad**
**Before creating a pull request make sure that:**

- [ ] commit messages are meaningful and follow good commit message guidelines
- [ ] README and other documentation has been updated / added (if needed)
- [ ] tests have been updated / new tests has been added (if needed)

Please remove this line and everything above and fill the following sections:

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ ] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/cmc/claimstore/documents/ClaimIssueReceiptService.java [hunk 1]
```diff
@@ -35,4 +36,9 @@ public ClaimIssueReceiptService(
         );
     }
 
+    @Override
+    public String filename(Claim claim) {
+        return buildClaimIssueReceiptFileBaseName(claim.getReferenceNumber());
```
Comment (sabahirfan) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/documents/ClaimIssueReceiptService.java]: are we missing file extension in file name here?

File: src/main/java/uk/gov/hmcts/cmc/claimstore/documents/DefendantResponseReceiptService.java [hunk 2]
```diff
@@ -29,8 +31,16 @@ public DefendantResponseReceiptService(
...
+            throw new NotFoundException("Defendant response does not exist for this claim");
+        }
         return pdfServiceClient.generateFromHtml(
             documentTemplates.getDefendantResponseReceipt(),
             contentProvider.createContent(claim));
     }
+
+    @Override
+    public String filename(Claim claim) {
+        return buildResponseFileBaseName(claim.getReferenceNumber());
```
Comment (sabahirfan) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/documents/DefendantResponseReceiptService.java]: Are we missing file extension here?

File: src/main/java/uk/gov/hmcts/cmc/claimstore/documents/SettlementAgreementCopyService.java [hunk 3]
```diff
@@ -29,11 +31,19 @@ public SettlementAgreementCopyService(
...
         byte[] settlementAgreement = documentTemplates.getSettlementAgreement();
         return pdfServiceClient.generateFromHtml(
             settlementAgreement,
             contentProvider.createContent(claim)
         );
     }
 
+    @Override
+    public String filename(Claim claim) {
+        return buildSettlementReachedFileBaseName(claim.getReferenceNumber());
```
Comment (sabahirfan) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/documents/SettlementAgreementCopyService.java]: same here, file extension

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/document/AlwaysGenerateDocumentsService.java [hunk 4]
```diff
@@ -41,25 +42,20 @@ public AlwaysGenerateDocumentsService(
...
+            case CLAIM_ISSUE_RECEIPT:
+                return claimIssueReceiptService.createPdf(getClaimByExternalId(externalId, authorisation));
+            case SEALED_CLAIM:
+                return sealedClaimPdfService.createPdf(getClaimByExternalId(externalId, authorisation));
+            case DEFENDANT_RESPONSE_RECEIPT:
+                return defendantResponseReceiptService.createPdf(getClaimByExternalId(externalId, authorisation));
+            case SETTLEMENT_AGREEMENT:
+                return settlementAgreementCopyService.createPdf(getClaimByExternalId(externalId, authorisation));
+        }
+        throw new RuntimeException(); //todo proper exception
```
Comment (sabahirfan) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/document/AlwaysGenerateDocumentsService.java]: How about adding a default case a throw IllegalArgumentException for `claimDocumentType`?

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/document/DocumentManagementBackedDocumentsService.java [hunk 5]
```diff
@@ -75,66 +65,42 @@ public DocumentManagementBackedDocumentsService(
...
+                return defendantResponseReceiptService;
+            case SETTLEMENT_AGREEMENT:
+                return settlementAgreementCopyService;
         }
-        return processRequest(claim,
-            authorisation,
-            DEFENDANT_RESPONSE_RECEIPT,
-            defendantResponseReceiptService,
-            buildResponseFileBaseName(claim.getReferenceNumber()));
+        throw new RuntimeException(); //todo proper exception
```
Comment (sabahirfan) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/document/DocumentManagementBackedDocumentsService.java]: same here

File: src/main/java/uk/gov/hmcts/cmc/claimstore/controllers/DocumentsController.java [hunk 6]
```diff
@@ -28,108 +29,21 @@
...
-    @ApiOperation("Returns a sealed claim copy for a given claim external id")
+    
+    @ApiOperation("Returns a specific pdf for a given claim external id")
     @GetMapping(
-        value = "/sealedClaim/{externalId}",
+        value = "/{documentType}/{externalId}",
         produces = MediaType.APPLICATION_PDF_VALUE
     )
-    public ResponseEntity<ByteArrayResource> sealedClaim(
+    public ResponseEntity<ByteArrayResource> document(
```
Comment (sabahirfan) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/controllers/DocumentsController.java]: We have to add tests for this as well.

File: src/main/java/uk/gov/hmcts/cmc/claimstore/documents/ClaimIssueReceiptService.java (hunk 1 above)
Comment (simad) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/documents/ClaimIssueReceiptService.java]: no, in these ones we append the extension when we create the PDF (doing `new PDF(bla)` appends the extension automatically; I find these different approaches a bit confusing

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/document/AlwaysGenerateDocumentsService.java (hunk 4 above)
Comment (simad) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/services/document/AlwaysGenerateDocumentsService.java]: yup

File: src/main/java/uk/gov/hmcts/cmc/claimstore/controllers/DocumentsController.java (hunk 6 above)
Comment (simad) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/controllers/DocumentsController.java]: yup

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/document/AlwaysGenerateDocumentsService.java [hunk 7]
```diff
@@ -41,25 +42,26 @@ public AlwaysGenerateDocumentsService(
     }
 
     @Override
-    public byte[] generateClaimIssueReceipt(String externalId, String authorisation) {
-        return claimIssueReceiptService.createPdf(getClaimByExternalId(externalId, authorisation));
+    public byte[] generateDocument(String externalId, ClaimDocumentType claimDocumentType, String authorisation) {
+        switch (claimDocumentType) {
+            case CLAIM_ISSUE_RECEIPT:
+                return claimIssueReceiptService.createPdf(
+                    getClaimByExternalId(externalId, authorisation)).getBytes();
```
Comment (karthikkannankainos) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/document/AlwaysGenerateDocumentsService.java]: getClaimByExternalId(externalId, authorisation) is common, so can be moved as single statement before switch

File: src/main/java/uk/gov/hmcts/cmc/claimstore/controllers/DocumentsController.java [hunk 8]
```diff
@@ -29,107 +30,20 @@ public DocumentsController(DocumentsService documentsService) {
...
     )
-    public ResponseEntity<ByteArrayResource> defendantResponseCopy(
+    public ResponseEntity<ByteArrayResource> document(
+        @ApiParam("Claim document type")
+        @PathVariable("documentType") @NotBlank String documentType,
         @ApiParam("Claim external id")
         @PathVariable("externalId") @NotBlank String externalId,
         @RequestHeader(HttpHeaders.AUTHORIZATION) String authorisation
     ) {
-        byte[] pdfDocument = documentsService.generateDefendantResponseReceipt(externalId, authorisation);
```
Comment (ckalista) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/controllers/DocumentsController.java]: Good weed out.

File: src/main/java/uk/gov/hmcts/cmc/claimstore/controllers/support/SupportController.java [hunk 9]
```diff
@@ -143,22 +143,8 @@ public void uploadDocumentToDocumentManagement(
...
-            case DEFENDANT_RESPONSE_RECEIPT:
-                documentsService.generateDefendantResponseReceipt(claim.getExternalId(), authorisation);
-                break;
-            case SETTLEMENT_AGREEMENT:
-                documentsService.generateSettlementAgreement(claim.getExternalId(), authorisation);
-                break;
-            default:
-                throw new BadRequestException("ClaimDocumentType " + claimDocumentType + " is not supported");
-        }
+        documentsService.generateDocument(claim.getExternalId(), claimDocumentType, authorisation);
```
Comment (ckalista) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/controllers/support/SupportController.java]: Much cleaner therefore!! 👍

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/document/AlwaysGenerateDocumentsService.java [hunk 10]
```diff
@@ -41,25 +42,26 @@ public AlwaysGenerateDocumentsService(
     }
 
     @Override
-    public byte[] generateClaimIssueReceipt(String externalId, String authorisation) {
-        return claimIssueReceiptService.createPdf(getClaimByExternalId(externalId, authorisation));
+    public byte[] generateDocument(String externalId, ClaimDocumentType claimDocumentType, String authorisation) {
```
Comment (ckalista) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/document/AlwaysGenerateDocumentsService.java]: 👍

File: domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/ClaimDocumentType.java [hunk 11]
```diff
@@ -1,12 +1,33 @@
...
+
+    public List<String> getValues() {
+        return values;
+    }
+
+    public static ClaimDocumentType fromValue(String value) {
+        return Arrays.stream(values())
+            .filter(v -> v.values.contains(value) || v.name().equals(value))
+            .findFirst()
+            .orElseThrow(() -> new RuntimeException("Unknown Claim Document Type: " + value));
```
Comment (sabahirfan) [inline domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/ClaimDocumentType.java]: `IllegalArgumentException` would be better option here

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/document/AlwaysGenerateDocumentsService.java (hunk 7 above)
Comment (simad) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/services/document/AlwaysGenerateDocumentsService.java]: 👍

File: domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/Claim.java [hunk 12]
```diff
@@ -147,16 +146,9 @@ public Claim(
...
-        if (claimDocumentCollection == null) {
-            return Optional.empty();
-        } else {
-            Optional<ClaimDocument> claimDocument = claimDocumentCollection.getDocument(claimDocumentType);
-            if (claimDocument.isPresent()) {
-                return Optional.ofNullable(claimDocument.get().getDocumentManagementUrl());
-            }
-        }
-        return Optional.empty();
+    public Optional<ClaimDocument> getClaimDocument(ClaimDocumentType claimDocumentType) {
```
[the remark this answers is not in the document]
Comment (simad) [reply domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/Claim.java]: hahahah true! 👍 I'll add one if it's not covered by something else, I'll check

**Pull request merged**